File: src/types/type.ts
import dayjs, { Dayjs, ManipulateType } from "dayjs";

export interface QueryTableDataParams {
    catalog: string;
    database: string;
    table: string;
    cluster: string;
    startDate: string;
    endDate: string;
    sort: string;
    timeField: string;
    data_filters: any[];
    interval?: IntervalEnum;
    interval_value?: number;
    search_type: string;
    search_value: string;
    indexes?: string;
    indexes_statement?: string;
    page: string;
    page_size: number;
    lucene_where?: string;
}


export interface SurroundingParams {
    catalog: string;
    database: string;
    table: string;
    cluster: string;
    timeField: string;
    time: string;
    data_filters: any;
    pageSize: string;
    operator: '>' | '<';
    theme: string | undefined;
}

export interface DataFilterType {
    fieldName: string;
    operator: Operator;
    value: Array<string | number>;
    label?: string;
    id: string;
}

export type Operator =
    | '='
    | '!='
    | 'in'
    | 'not in'
    | 'is null'
    | 'is not null'
    | 'between'
    | 'not between'
    | 'like'
    | 'not like'
    | 'match_all'
    | 'match_any'
    | 'match_phrase'
    | 'match_phrase_prefix';

export interface DiscoverCurrent {
    catalog: string;
    database: string;
    table: string;
    cluster: string;
    timeField: string;
    date: [Dayjs, Dayjs] | [];
}

export enum IntervalEnum {
    Auto = 'auto',
    Day = 'day',
    Week = 'week',
    Month = 'month',
    Year = 'year',
    Hour = 'hour',
    Minute = 'minute',
    Second = 'second',
}


export type ShortcutItem = {
    key: string;
    text: string;
    label: string;
    range: (now?: dayjs.Dayjs) => dayjs.Dayjs[];
    format: string;
    type: string;
    number: number;
}

export interface QueryTableFieldsParams {
    catalog: string;
    database: string;
    table: string;
    describe_extend_variant_column?: string;
}

export type FieldNode = {
    name: string;
    type: string;
    children?: FieldNode[];
    Null?: string;
    Key?: string;
    Default?: string | null;
    Extra?: string;
    Comment?: string;
    Visible?: string;
};


export interface RequestParams {
    catalog: string;
    database: string;
    table: string;
    describe_extend_variant_column?: boolean;
}

export interface QueryTableIndexesParams {
    catalog: string;
    database: string;
    table: string
}

export interface QueryTableParams {
    catalog: string;
    database: string;
}

export type AutoInterval = {
    current_date_range: number;
    current_type: ManipulateType;
    interval_unit: ManipulateType;
    interval_value: number;
};


export type Frame = {
    schema: {
        fields: Array<{ name: string }>;
    };
    data: {
        values: any[][];
    };
};

export type Result = {
    frames: Frame[];
};

export type Root = {
    results: Record<string, Result>;
};

export type QueryTracesParams = {
    database: string;
    table: string;
    timeField: string;
    startDate: string;
    endDate: string;
    service_name?: string;
    operation?: string;
    statusCode?: string;
    minDuration?: string;
    maxDuration?: string;
    tags?: string;
    page?: number;
    page_size?: number;
    sortBy?: string;
};

export type TracesServicesParams = {
  table: string;
  timeField: string;
  startDate: string;
  endDate: string;
};

export type TracesOperationsParams = {
  table: string;
  timeField: string;
  startDate: string;
  endDate: string;
  service_name: string;
};

File: pkg/plugin/type.go
package plugin

type DataFilter struct {
	// 自定义结构体，你需要根据实际结构定义字段，比如：
	FieldName string
	Operator  string
	Value     []interface{}
}

type QueryTraceDetailParams struct {
	DS       string // 数据源标识
	Catalog  string `json:"catalog"`
	Database string `json:"database"`
	Table    string `json:"table"`
	Cluster  string `json:"cluster"`
	TraceId  string `json:"trace_id"`
}

type QueryTableDataParams struct {
	Catalog          string       `json:"catalog"`
	Database         string       `json:"database"`
	Table            string       `json:"table"`
	Cluster          string       `json:"cluster"`
	StartDate        string       `json:"startDate"`
	EndDate          string       `json:"endDate"`
	Sort             string       `json:"sort"`
	TimeField        string       `json:"timeField"`
	Interval         string       `json:"interval"`       // 可用枚举限制
	IntervalValue    int          `json:"interval_value"` // 数值，需转为 int
	DataFilters      []DataFilter `json:"data_filters"`   // 自定义类型数组，需提前定义
	SearchType       string       `json:"search_type"`    // 'Search' | 'SQL'
	SearchValue      string       `json:"search_value"`
	Indexes          string       `json:"indexes"`
	IndexesStatement string       `json:"indexes_statement"` // 用于 Search 模式的索引查询
	DS               string       // 数据源标识
	PageSize         int          `json:"page_size"` // 分页大小
	Page             int          `json:"page"`      // 当前页码
	TraceId          string       `json:"trace_id"`
}

// type QueryTracesParams struct {
// 	Catalog   string `json:"catalog"`
// 	Database  string `json:"database"`
// 	Table     string `json:"table"`
// 	Cluster   string `json:"cluster"`
// 	StartDate string `json:"startDate"`
// 	EndDate   string `json:"endDate"`
// 	TimeField string `json:"timeField"`
// 	DS        string // 数据源标识
// 	PageSize  int    `json:"page_size"` // 分页大小
// 	Page      int    `json:"page"`      // 当前页码
// }

type SurroundingParams struct {
	Table       string
	Database    string
	TimeField   string
	Operator    string // "<" or ">"
	Time        string
	DataFilters []DataFilter `json:"data_filters"`
	PageSize    int
	DS          string // 数据源标识
}

type TracesServicesParams struct {
	Table     string
	Database  string
	TimeField string
	StartDate string `json:"startDate"`
	EndDate   string `json:"endDate"`
	DS        string // 数据源标识
}

type TracesOperationsParams struct {
	Table       string
	Database    string
	TimeField   string
	ServiceName string `json:"serviceName"`
	StartDate   string `json:"startDate"`
	EndDate     string `json:"endDate"`
	DS          string // 数据源标识
}

type RawTraceRow struct {
	TraceID         string `json:"trace_id"`
	RootServiceName string `json:"root_service_name"`
	RootSpanName    string `json:"root_span_name"`
	RootSpanKind    string `json:"root_span_kind"`
	RootStatusCode  string `json:"root_status_code"`
	Operation       string `json:"operation"`
	ResourceAttrs   string `json:"resource_attributes"`
	StartTime       string `json:"start_time"` // 改成 string
	EndTime         string `json:"end_time"`   // 改成 string
	DurationUs      int64  `json:"duration_us"`
	SpanCount       int64  `json:"span_count"`
}

type QueryTracesParams struct {
	Catalog        string `json:"catalog"`
	Database       string `json:"database"`
	Table          string `json:"table"`
	Cluster        string `json:"cluster"`
	StartDate      string `json:"startDate"`
	EndDate        string `json:"endDate"`
	TimeField      string `json:"timeField"`
	DS             string // 数据源标识
	PageSize       int    `json:"page_size"` // 分页大小
	Page           int    `json:"page"`      // 当前页码
	ServiceName    string `json:"service_name"`
	Operation      string `json:"operation"`
	StatusCode     string `json:"status_code"`
	MinDurationStr string `json:"minDuration"` // 单位微秒
	// 例如：1000us, 1s, 500ms
	MaxDurationStr string `json:"maxDuration"`
	Tags           string // logfmt 格式：error=true http.url="http://abc.com"
	SortBy         string `json:"sort_by"` // 'most-recent' | 'longest-duration'
}

type AggregatedTrace struct {
	TraceID         string
	RootServiceName string
	RootSpanName    string
	RootSpanKind    string
	RootStatusCode  string
	Operation       string
	ServiceTags     string // 经过聚合转换后的字符串
	StartTime       string
	EndTime         string
	DurationUs      int64
	SpanCount       int64
}
